src/ApiCompat/Microsoft.DotNet.ApiCompatibility/Rules/AttributesMustMatch.cs

```diff
@@ -0,0 +1,263 @@
+        }
+
+        private void RunOnTypeSymbol(ITypeSymbol? left, ITypeSymbol? right, string leftName, string rightName, IList<CompatDifference> differences)
+        {
+            if (left is null || right is null)
+            {
+                return;
+            }
+            if (left is INamedTypeSymbol leftNamed && right is INamedTypeSymbol rightNamed)
+            {
+                if (leftNamed.TypeParameters.Length == rightNamed.TypeParameters.Length)
+                {
+                    for (int i = 0; i < leftNamed.TypeParameters.Length; i++)
+                    {
+                        reportAttributeDifferences(left, left.GetDocumentationCommentId() + $"<{i}>", leftNamed.TypeParameters[i].GetAttributes(), rightNamed.TypeParameters[i].GetAttributes(), differences);
+                    }
+                }
+            }
+            reportAttributeDifferences(left, left.GetDocumentationCommentId() ?? "", left.GetAttributes(), right.GetAttributes(), differences);
+        }
+
+        private static CompatDifference removedDifference(ISymbol containing, string itemRef, AttributeData? attr)
+        {
+            string msg = string.Format(Resources.CannotRemoveAttribute, attr, containing);
+            return new CompatDifference(DiagnosticIds.CannotRemoveAttribute, msg, DifferenceType.Removed, itemRef + ":[" + attr?.AttributeClass?.GetDocumentationCommentId() + "]");
+        }
+
+        private static CompatDifference addedDifference(ISymbol containing, string itemRef, AttributeData? attr)
+        {
+            string msg = string.Format(Resources.CannotAddAttribute, attr, containing);
+            return new CompatDifference(DiagnosticIds.CannotAddAttribute, msg, DifferenceType.Added, itemRef + ":[" + attr?.AttributeClass?.GetDocumentationCommentId() + "]");
+        }
+
+        private static CompatDifference changedDifference(ISymbol containing, string itemRef, AttributeData? attr)
+        {
+            string msg = string.Format(Resources.CannotChangeAttribute, attr?.AttributeClass, containing);
+            return new CompatDifference(DiagnosticIds.CannotChangeAttribute, msg, DifferenceType.Changed, itemRef + ":[" + attr?.AttributeClass?.GetDocumentationCommentId() + "]");
+        }
+
+        private bool attributeEquals(AttributeData? left, AttributeData? right)
+        {
+            if (left != null && right != null)
+            {
+                if (!_settings.SymbolComparer.Equals(left.AttributeClass!, right.AttributeClass!))
+                {
+                    return false;
+                }
+                if (!Enumerable.SequenceEqual(left.ConstructorArguments, right.ConstructorArguments))
+                {
+                    return false;
+                }
+                return Enumerable.SequenceEqual(left.NamedArguments, right.NamedArguments);
+            }
+            return left == right;
+        }
+
+        private void reportAttributeDifferences(ISymbol containing,
```

Comment:
In general, all methods are capital cased in C#, including private ones, i.e. https://github.com/dotnet/runtime/blob/e71a9583b4d6c9bd97edd87cda7f98f232f63530/src/libraries/System.Private.CoreLib/src/System/String.cs#L66.